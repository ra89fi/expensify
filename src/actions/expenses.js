import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const { description = '', notes = '', amount = 0, createdAt = Date.now() } = expenseData;
    const expense = { description, notes, amount, createdAt };
    return database
      .ref('expenses')
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      })
      .catch(err => console.log(err.message));
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});
