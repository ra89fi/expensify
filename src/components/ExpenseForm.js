import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import CurrencyDropdown from './CurrencyDropdown';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      notes: props.expense ? props.expense.notes : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      currency: props.expense ? props.expense.currency : 'USD',
      focused: null,
      error: ''
    };
  }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  };
  onNotesChange = e => {
    const notes = e.target.value;
    this.setState(() => ({
      notes
    }));
  };
  onCurrencyChange = e => {
    const currency = e.target.value;
    this.setState(() => ({
      currency
    }));
  };
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount
      }));
    }
  };
  onDateChange = createdAt => {
    if (!createdAt) return;
    this.setState(() => ({
      createdAt
    }));
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      focused
    }));
  };
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: '* Please provide description and amount !'
      }));
    } else {
      this.setState(() => ({
        error: ''
      }));
      this.props.onSubmit({
        description: this.state.description,
        notes: this.state.notes,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        currency: this.state.currency
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Description"
          className="text-input"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <div className="form__money">
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <input
            type="text"
            className="text-input"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <CurrencyDropdown
            cName="select-input"
            currency={this.state.currency}
            onCurrencyChange={this.onCurrencyChange}
          />
        </div>
        <textarea
          className="textarea-input"
          cols="20"
          rows="5"
          placeholder="Add a note for your expense (optional)"
          value={this.state.notes}
          onChange={this.onNotesChange}
        />
        <div>
          <button className="button">Save Expense</button>
        </div>
      </form>
    );
  }
}
