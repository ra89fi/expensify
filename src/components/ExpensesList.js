import React from 'react';
import { connect } from 'react-redux';
import ExpensesListItem from './ExpensesListItem';
import selectExpenses from '../selectors/expenses';

export const ExpensesList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="visible-mobile">Expenses</div>
      <div className="visible-desktop">Expense</div>
      <div className="visible-desktop">Amount</div>
    </div>
    <div className="list-body">
      {props.expenses.length === 0 ? (
        <div className="list-item list-item__message">
          <span>No expenses</span>
        </div>
      ) : (
        props.expenses.map(expense => <ExpensesListItem key={expense.id} {...expense} />)
      )}
    </div>
  </div>
);

const ConnectedExpensesList = connect(state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
})(ExpensesList);

export default ConnectedExpensesList;
