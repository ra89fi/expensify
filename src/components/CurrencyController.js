import React from 'react';
import { connect } from 'react-redux';
import CurrencyDropdown from './CurrencyDropdown';
import { setCurrency } from '../actions/settings';

export class CurrencyController extends React.Component {
  onCurrencyChange = e => {
    this.props.setCurrency(e.target.value);
  };
  render() {
    return (
      <div className="currency">
        <span className="currency__title">Show as &nbsp;: &nbsp;</span>
        <CurrencyDropdown
          cName="currency__select"
          currency={this.props.currency}
          onCurrencyChange={this.onCurrencyChange}
        />
      </div>
    );
  }
}

export default connect(
  state => ({
    currency: state.settings.currency
  }),
  dispatch => ({
    setCurrency: currency => dispatch(setCurrency(currency))
  })
)(CurrencyController);
