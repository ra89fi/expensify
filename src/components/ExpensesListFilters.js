import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, setSortBy, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    focused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = focused => {
    this.setState(() => ({
      focused
    }));
  };
  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };
  onSelectChange = e => {
    this.props.setSortBy(e.target.value);
  };
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              placeholder="Search expenses"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select-input"
              name="filters"
              id="filters"
              value={this.props.filters.sortBy}
              onChange={this.onSelectChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.focused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

export default connect(
  state => ({
    filters: state.filters
  }),
  dispatch => ({
    setTextFilter: text => dispatch(setTextFilter(text)),
    setSortBy: option => dispatch(setSortBy(option)),
    setStartDate: date => dispatch(setStartDate(date)),
    setEndDate: date => dispatch(setEndDate(date))
  })
)(ExpenseListFilters);
