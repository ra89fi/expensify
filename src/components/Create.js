import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

const Create = props => (
  <div>
    <p>Add Expense</p>
    <ExpenseForm
      onSubmit={expense => {
        props.dispatch(addExpense(expense));
        props.history.push('/');
      }}
    />
  </div>
);

export default connect()(Create);
