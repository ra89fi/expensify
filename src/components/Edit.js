import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class Edit extends React.Component {
  onSubmit = expense => {
    this.props.onSubmit(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onBtnClick = () => {
    this.props.onBtnClick({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Edit Expense</h2>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button button__remove" onClick={this.onBtnClick}>
            Remove Expense
          </button>
        </div>
      </div>
    );
  }
}

export default connect(
  (state, props) => {
    return {
      expense: state.expenses.find(expense => expense.id == props.match.params.id)
    };
  },
  (dispatch, props) => {
    return {
      onSubmit: (id, expense) => dispatch(startEditExpense(id, expense)),
      onBtnClick: data => dispatch(startRemoveExpense(data))
    };
  }
)(Edit);
