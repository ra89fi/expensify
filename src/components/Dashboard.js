import React from 'react';
import ExpensesList from './ExpensesList';
import ExpenseListFilters from './ExpensesListFilters';
import ExpensesSummary from './ExpensesSummary';

const Dashboard = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpensesList />
  </div>
);

export default Dashboard;
