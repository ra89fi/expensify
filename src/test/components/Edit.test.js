import React from 'react';
import { shallow } from 'enzyme';
import { Edit } from '../../components/Edit';
import expenses from '../fixtures/expenses';

let history, wrapper, onSubmit, onBtnClick;

beforeEach(() => {
  history = {
    push: jest.fn()
  };
  onSubmit = jest.fn();
  onBtnClick = jest.fn();
  wrapper = shallow(
    <Edit expense={expenses[0]} onSubmit={onSubmit} onBtnClick={onBtnClick} history={history} />
  );
});

test('should render Edit correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(onSubmit).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test('should handle remove expense', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(onBtnClick).toHaveBeenLastCalledWith({ id: expenses[0].id });
});
