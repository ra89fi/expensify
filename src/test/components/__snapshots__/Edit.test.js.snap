// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Edit correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        Edit Expense
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "Gum",
          "id": "1",
          "notes": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button__remove"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
